### PR TITLE
completed prime numbers challenge

### LINKED ISSUE #32: Checking Prime Numbers
Write a function that checks if the given number is prime.
A prime number is divisible only by 1 and itself.

### PR DESCRIPTION
**Pull request by Alaa-Elgozouli**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by Alaa-Elgozouli** [comment]
Hi, I checked to see if my code passes all CI checks, however, it did not pass the "py tests". I know what the problem is but I have no idea how to solve it. Would appreciate the feedback!

**Comment by Tbaosman** [comment]
> Hi, I checked to see if my code passes all CI checks, however, it did not pass the "py tests". I know what the problem is but I have no idea how to solve it. Would appreciate the feedback!

To pass the tests; your function name must match the function you're trying to import in the test file. In your case, your function name is is_prime(), and in the test file you're trying to import prime_numbers. we are importing the functions, not the files. 
so, just change the function name in the function file from **is_prime() to prime_numbers()** (which is actually a check in the checklist stating that your function name must be the same as file's name)

File: solutions/prime_numbers.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+Module contents:
+    - is_prime: checks if the given number is a prime number.
+
+created on 04/01/2025
+@author: Alaa Mohamed
+
+"""
+
+
+def is_prime(N: int) -> bool:
```
Comment (Tbaosman) [inline solutions/prime_numbers.py]: change the function name to make it match the file name and also change the function parameter

File: solutions/tests/test_prime_numbers.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+def test_not_int(self):
+    """It should raise AssertionError for non-integer input"""
+    with self.assertRaises(AssertionError):
+        prime_numbers(1.5)
+
+
+def test_edge_case(self):
+    """Testing the function with edge inputs"""
+    self.assertFalse(prime_numbers(1))
+    self.assertFalse(prime_numbers(0))
```
Comment (Tbaosman) [inline solutions/tests/test_prime_numbers.py]: Only one assertion must be in a single unit test

**Review by Tbaosman — CHANGES_REQUESTED** [review]

File: solutions/prime_numbers.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    Returns -> bool: True if the number is prime, False otherwise.
+
+    Raises:
+        AssertionError: if the input is not an integer.
+
+    Examples:
+    >>> prime_numbers("1")
+    False
+    >>> prime_numbers("2")
```
Comment (Tbaosman) [inline solutions/prime_numbers.py]: input in the docstring tests should be of the type "int"

File: solutions/prime_numbers.py (hunk 1 above)
Comment (Alaa-Elgozouli) [reply solutions/prime_numbers.py]: Done, thanks!!

File: solutions/prime_numbers.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    True
+    """
+    assert isinstance(number, int), "input must be an integer"
+
+    if number <= 1:
+        return False
+    for i in range(2, number):
+        if number % i == 0:
+            return False
+        return True
```
Comment (Tbaosman) [inline solutions/prime_numbers.py]: There is a big **logic flaw** in line 45; the return statement is too early in your implementation which will lead to errors in identifying some prime number. for example, for the number (2), -by this current logic- the return will be `none` as it can't even be checked. Also number (9), the function will mistakenly return `true` while it's not a prime number.
it's just an indentation error so double check this line again.

**Review by Tbaosman — CHANGES_REQUESTED** [review]

File: solutions/prime_numbers.py (hunk 3 above)
Comment (Alaa-Elgozouli) [reply solutions/prime_numbers.py:33]: aren't 1 & 2 integers? not sure I got your point

**Review by Tbaosman — APPROVED** [review]
Test cases are "fine", but I'd suggest adding more to ensure robustness. Still, I'll approve it. Good work!
